sharding/notary/notary.go

```diff
@@ -151,7 +205,287 @@ func joinNotaryPool(client mainchain.Client) error {
+func submitVote(shard sharding.Shard, client mainchain.Client, headerHash *common.Hash) error {
+
+	shardID := shard.ShardID()
+	// checks if the shardID is valid
+	if shardID == big.NewInt(0) || shardID.Int64() > sharding.ShardCount {
+		return fmt.Errorf("ShardId is invalid, it has to be between %d and %d, instead it is %v", 0, sharding.ShardCount, shardID)
+	}
+
+	currentblock, err := client.ChainReader().BlockByNumber(context.Background(), nil)
+	if err != nil {
+		return fmt.Errorf("Unable to retrieve last block: %v", err)
+	}
+
+	period := big.NewInt(0).Div(currentblock.Number(), big.NewInt(sharding.PeriodLength))
+
+	collPeriod, err := client.SMCCaller().LastSubmittedCollation(&bind.CallOpts{}, shardID)
+
+	if err != nil {
+		return fmt.Errorf("Unable to get period from last submitted collation: %v", err)
+	}
+
+	// Checks if the current period is valid in order to vote for the collation on the shard
+	if period.Int64() != collPeriod.Int64() {
+		return fmt.Errorf("Period in collation is not equal to current period : %d , %d", collPeriod, period)
+	}
+
+	nreg, err := getNotaryRegistry(client)
+
+	if err != nil {
+		return err
+	}
+
+	if !nreg.Deposited {
+		return fmt.Errorf("Notary has not deposited to the SMC")
+	}
+
+	// Checking if the pool index is valid
+	if nreg.PoolIndex.Int64() >= sharding.NotaryCommitSize {
+		return fmt.Errorf("Invalid Pool Index %d as it is more than the committee size of %d", nreg.PoolIndex, sharding.NotaryCommitSize)
+	}
+
+	collationRecords, err := client.SMCCaller().CollationRecords(&bind.CallOpts{}, shardID, period)
+
+	if err != nil {
+		return fmt.Errorf("Unable to get collation record: %v", err)
+	}
+
+	chunkroot, err := shard.ChunkRootfromHeaderHash(headerHash)
+
+	if err != nil {
+		return fmt.Errorf("Unable to get chunk root: %v", err)
+	}
+
+	// Checking if the chunkroot is valid
+	if !bytes.Equal(collationRecords.ChunkRoot[:], chunkroot.Bytes()) {
+		return fmt.Errorf("Submmitted collation header has a different chunkroot to the one saved in the SMC")
+
+	}
+
+	hasVoted, err := client.SMCCaller().HasVoted(&bind.CallOpts{}, shardID, nreg.PoolIndex)
+
+	if err != nil {
+		return fmt.Errorf("Unable to know if notary voted: %v", err)
+	}
+
+	if hasVoted {
+		return errors.New("Notary has already voted")
+	}
+
+	inCommitee, err := client.SMCCaller().GetNotaryInCommittee(&bind.CallOpts{}, shardID)
+
+	if err != nil {
+		return fmt.Errorf("Unable to know if notary is in committee: %v", err)
+	}
+
+	if inCommitee != client.Account().Address {
+		return errors.New("Notary is not eligible to vote in this shard at the current period")
+	}
+
+	txOps, err := client.CreateTXOpts(nil)
+	if err != nil {
+		return fmt.Errorf("unable to create txOpts: %v", err)
+	}
+
+	tx, err := client.SMCTransactor().SubmitVote(txOps, shardID, period, nreg.PoolIndex, collationRecords.ChunkRoot)
+
+	if err != nil {
+
+		return fmt.Errorf("Unable to submit Vote: %v", err)
+	}
+
+	receipt, err := client.TransactionReceipt(tx.Hash())
```

Comment:
Is this supposed to wait for the transaction to be mined? There's no way that you've received a receipt by now.